src/Lecture4.hs

```diff
@@ -157,6 +176,19 @@ string.
 If both strings have the same length, return the first one.
 -}
 instance Semigroup MaxLen where
+    (<>) :: MaxLen -> MaxLen -> MaxLen
+    a <> b = MaxLen (longest (unMaxLen a) (unMaxLen b))
+
+longest :: [a] -> [a] -> [a]
+longest l1 l2 = case longestEither l1 l2 of
+    Left _ -> l1
+    Right _ -> l2
+
+longestEither :: [a] -> [a] -> Either () ()
+longestEither l1 l2 = case (l1, l2) of
+    (_, []) -> Left ()
+    ([], _) -> Right ()
+    (_:xs, _:ys) -> longestEither xs ys
```

Comment:
That's a clever implementation of finding the longest list 🧠 🏆 
It works in time proportional to the length of shortest list.

You can improve the performance even further. In the `MaxLen` data type you can store the length of the string as well (so it'll be a two-field data type). And you can compare by lengths even faster